src/internal/plugins/controllers/alert-flash-content.ts

```diff
@@ -0,0 +1,91 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// this code should not depend on React typings, because it is portable between major versions
+interface RefShim<T> {
+  current: T | null;
+}
+
+export interface AlertFlashContentContext {
+  type: string;
+  headerRef: RefShim<HTMLElement>;
+  contentRef: RefShim<HTMLElement>;
+  actionsRef: RefShim<HTMLElement>;
+}
+
+export type ReplacementType = 'original' | 'remove' | ((container: HTMLElement) => void);
+export type ReplacementTypeSafe = 'original' | 'remove' | true;
+
+type RegisterReplacement = (type: 'header' | 'content', doReplacement: ReplacementType) => void;
+
+export interface AlertFlashContentResult {
+  update: () => void;
+  unmount: (containers: { replacementHeaderContainer: HTMLElement; replacementContentContainer: HTMLElement }) => void;
+}
+
+export interface AlertFlashContentConfig {
+  id: string;
+  runReplacer: (context: AlertFlashContentContext, registerReplacement: RegisterReplacement) => AlertFlashContentResult;
+}
+
+export interface AlertFlashContentRegistrationListener {
+  (provider?: AlertFlashContentConfig): void | (() => void);
+  cleanup?: void | (() => void);
+}
+
+export interface AlertFlashContentApiPublic {
+  registerContentReplacer(config: AlertFlashContentConfig): void;
+}
+
+export interface AlertFlashContentApiInternal {
+  clearRegisteredReplacer(): void;
+  onContentRegistered(listener: AlertFlashContentRegistrationListener): () => void;
+}
+
+export class AlertFlashContentController {
+  #listeners: Array<AlertFlashContentRegistrationListener> = [];
+  #cleanups = new Map<AlertFlashContentRegistrationListener, null | (() => void)>();
+  #provider?: AlertFlashContentConfig;
+
+  registerContentReplacer = (content: AlertFlashContentConfig) => {
+    if (this.#provider) {
+      console.warn(
+        `Cannot call \`registerContentReplacer\` with new provider: provider with id "${this.#provider.id}" already registered.`
+      );
+      return;
+    }
+    this.#provider = content;
+
+    // Notify existing components if registration happens after the components are rendered.
+    this.#listeners.forEach(listener => {
```

Comment:
This recent change turns the API from asynchronous to synchronous.

I see that you updated all tests to reflect that, but did you an analysis of pros and cons of this decision? What did you find?

Reply:
That affects our testing but does not make any difference to the public API. I removed the debounce because there is no evidence it is actually needed. When content changes - only one listener is updated. We only update all when the registerContentReplacer is called which is expected to be called once. It can technically happen while there are a few alerts/flash items rendered but I don't anticipate any issues when triggering a few state updates in a loop this way.

Reply:
It is a breaking change. Having it async from the start (even with a dummy `await Promise.resolve()`) is future proof because it allows future changes without actions on the consumers end.

Converting it from sync to async in the future will be a breaking change.

Not sure if you can make this decision this way, just between two PR revisions. Some additional evaluation is needed.

Reply:
Updated